src/Mono.Android/Android.Runtime/AndroidRuntime.cs

```diff
@@ -351,41 +365,284 @@ public override void RegisterNativeMembers (JniType jniType, Type type, string m
 				natives [i] = new JniNativeMethodRegistration (toks [0], toks [1], callback);
 			}
 
-			JniEnvironment.Types.RegisterNatives (jniType.PeerReference, natives, natives.Length);
+			JniEnvironment.Types.RegisterNatives (nativeClass.PeerReference, natives, natives.Length);
 		}
 	}
 
 	class AndroidValueManager : JniRuntime.JniValueManager {
 
+		Dictionary<IntPtr, List<WeakReference<IJavaPeerable>>> instances = new Dictionary<IntPtr, List<WeakReference<IJavaPeerable>>> (new InstancesKeyComparer ());
```

Comment:
This shouldn't affect the performance (besides the added IJavaPeerable) as it is moved from Java.Lang.Object here, but makes me wonder whether we might  be able to improve performance by using some other "specialized" data structure to avoid allocating all the lists? Do not have any specific DS in mind yet, but looks like a good potential improvement to me.

Reply:
Introduced a new `IdentityHashTargets` type which contains a `WeakReference<IJavaPeerable>` field with a list fallback.  This way, if an identity has code only has one mapping, we avoid the `List<T>` allocation.

We could reasonably extend this to hold additional fields without hitting the `List<T>`, if necessary.